File: CarePackage/Software/Krita.cs
namespace CarePackage.Software;

public class Krita : BaseSoftware
{
    public override string           Key            { get; set; } = "krita";
    public override string           Name           { get; set; } = "Krita";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Creative;
    public override string           DownloadName   { get; set; } = "krita-x64-setup.exe";
    public override bool             IsArchive      { get; set; } = false;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = true;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.krita;
    public override string           Homepage       { get; set; } = "https://krita.org";

    private readonly HttpClient _http;

    public Krita(IServiceProvider services)
    {
        _http = services.GetKeyedService<HttpClient>("MimicBrowser")!;
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var res = await _http.GetAsync("https://krita.org/en/download", ct);
        
        res.EnsureSuccessStatusCode();

        var downloadUrlPattern = new Regex(@"https://download\.kde\.org/stable/krita/\d{1,}\.\d{1,}\.\d{1,}/krita-x64-\d{1,}\.\d{1,}\.\d{1,}-setup\.exe");
        var html               = await res.Content.ReadAsStringAsync(ct);
        var match              = downloadUrlPattern.Match(html);
        
        DownloadUrlResolveException.ThrowUnless(match.Success);

        return match.Groups[0].Value;
    }
}

File: packages/app/src/services/software/definitions/krita.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'krita';
	public name = 'Krita';
	public category = [SoftwareCategory.Creative];
	public downloadName = 'krita-x64-setup.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'krita.png';
	public homepage = 'https://krita.org';

	public async resolveDownloadUrl() {
		const res = await fetch('https://krita.org/en/download');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/download\.kde\.org\/stable\/krita\/\d+\.\d+\.\d+\/krita-x64-\d+\.\d+\.\d+-setup\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}
